add retries with exponential back-off for remote cache

tdyas opened the pull request:
## Problem

When using remote cache, transient errors returned by the remote system or local client stack (e.g, gRPC "unavailable" status) are not retried. This is not as good an experience for users of remote cache as they will see warnings about remote cache issues.

## Solution

Retry failed remote cache requests using exponential backoff. This PR introduces a helper called `retry_call` which implements the backoff logic and ports some of the remote cache/execution call sites to use it.

This PR does not port the CAS write logic since the use of streams there complicates usage of a function like `retry_call`.

Fixes https://github.com/pantsbuild/pants/issues/11373.

[ci skip-build-wheels]

tdyas commented:
Putting up to get general input on direction. I still need to audit more call sites but that can be done in parallel to getting your initial comments.

tdyas commented on src/rust/engine/fs/store/src/remote.rs:
```diff
@@ -363,12 +363,13 @@ impl ByteStore {
-        .await?;
+      let client = store2.cas_client.as_ref().clone();
+      let response = retry_call(client, move |mut client| {
+        let request = request.clone();
```
I do not like having to `clone` the request here, but it is required by Rust's ownership rules since the Tonic client consumes the request. Are there alternatives to this?

Eric-Arellano replied:
Ah, I found the issue! If you were to change `retry_call` to be `FnOnce`, then you can avoid the clone. But of course, it needs to be `Fn` because you call the function multiple times.

```diff
diff --git a/src/rust/engine/fs/store/src/remote.rs b/src/rust/engine/fs/store/src/remote.rs
index 43e5fbecb..d68ab59dc 100644
--- a/src/rust/engine/fs/store/src/remote.rs
+++ b/src/rust/engine/fs/store/src/remote.rs
@@ -365,7 +365,7 @@ impl ByteStore {
       let store2 = store.clone();
       let client = store2.cas_client.as_ref().clone();
       let response = retry_call(client, move |mut client| {
-        let request = request.clone();
+        // let request = request.clone();
         async move { client.find_missing_blobs(request).await }
       })
       .await
diff --git a/src/rust/engine/grpc_util/src/retry.rs b/src/rust/engine/grpc_util/src/retry.rs
index 60f817b83..8680be2f5 100644
--- a/src/rust/engine/grpc_util/src/retry.rs
+++ b/src/rust/engine/grpc_util/src/retry.rs
@@ -24,16 +24,17 @@ fn is_retryable(status: &Status) -> bool {
 pub async fn retry_call<T, C, F, Fut>(client: C, f: F) -> Result<Response<T>, Status>
 where
   C: Clone,
-  F: Fn(C) -> Fut,
+  F: FnOnce(C) -> Fut,
   Fut: Future<Output = Result<Response<T>, Status>>,
 {
   const INTERVAL_DURATION: Duration = Duration::from_millis(10);
   const MAX_RETRIES: u32 = 3;
   const MAX_BACKOFF_DURATION: Duration = Duration::from_secs(5);
 
-  let mut last_error: Option<Status> = None;
+  let last_error: Option<Status> = None;
 
   let mut num_retries = 0;
+  f(client);
   while num_retries < MAX_RETRIES {
     // Delay before the next send attempt if this is a retry.
     if num_retries > 0 {
@@ -43,18 +44,18 @@ where
       tokio::time::sleep(sleep_time).await;
     }
 
-    let client2 = client.clone();
-    let result_fut = f(client2);
-    match result_fut.await {
-      Ok(r) => return Ok(r),
-      Err(status) => {
-        if is_retryable(&status) {
-          last_error = Some(status);
-        } else {
-          return Err(status);
-        }
-      }
-    }
+    // let client2 = client.clone();
+    // let result_fut = f(client2);
+    // match result_fut.await {
+    //   Ok(r) => return Ok(r),
+    //   Err(status) => {
+    //     if is_retryable(&status) {
+    //       last_error = Some(status);
+    //     } else {
+    //       return Err(status);
+    //     }
+    //   }
+    // }
 
     num_retries += 1
   }
```

From `rustc --explain E0507`:

> This can also happen when using a type implementing `Fn` or `FnMut`, as neither
allows moving out of them (they usually represent closures which can be called
more than once). Much of the text following applies equally well to non-`FnOnce`
closure bodies.

illicitonion replied:
Yeah, because tonic takes ownership of the request, there's no improvement to be had here. In theory I can imagine `tonic`/`prost` allowing you to encode from an `Arc<Message>` or similar, but they don't, so... :) On the plus side, a request shouldn't be _too_ expensive to clone, right?

The clone of `client`, on the other hand, could theoretically be avoided... I think this really wants to be a `FnMut() -> Fut` rather than a `Fn(C) -> Fut` ("Can be called multiple times, but never more than once at a time"), but async closures are unstable and also kinda funky so I don't know of a way to _actually_ avoid the clone of `client`.

tdyas replied:
> The clone of client, on the other hand, could theoretically be avoided... I think this really wants to be a FnMut() -> Fut rather than a Fn(C) -> Fut ("Can be called multiple times, but never more than once at a time"), but async closures are unstable and also kinda funky so I don't know of a way to actually avoid the clone of client.

The client at least is intended to be cheap to Clone (via an Arc).

tdyas commented on src/rust/engine/grpc_util/src/retry.rs:
```diff
@@ -0,0 +1,63 @@
+  F: Fn(C) -> Fut,
+  Fut: Future<Output = Result<Response<T>, Status>>,
+{
+  const INTERVAL_DURATION: Duration = Duration::from_millis(10);
+  const MAX_RETRIES: u32 = 3;
+  const MAX_BACKOFF_DURATION: Duration = Duration::from_secs(5);
```
Should these be configurable?

Eric-Arellano replied:
I don't think so, at least to start. Imo there's a risk of options fatigue. For example, how is an end user supposed to determine what values make sense? These are finicky to play with.

Another important detail here: remote cache failures should not break Pants. If these numbers aren't perfectally optimal, that's not a big deal.

illicitonion replied:
I could buy `MAX_RETRIES` being configurable (people may want different trade-offs there), but I can't really justify the others.

tdyas replied:
I'll keep them non-configurable for now.

tdyas commented on src/rust/engine/grpc_util/src/retry.rs:
```diff
@@ -0,0 +1,63 @@
+    num_retries += 1
+  }
+
+  Err(last_error.unwrap())
```
This should be safe as it will always be set by line 52, but would love a second opinion.

Eric-Arellano replied:
Yes, I agree this is safe. However, could you please change it to be `.expect()` and say something like "remote cache retry logic did not set `last_error` properly! Please file an issue at https://github.com/pantsbuild/pants/issues"

I'm thinking about that current failure we have "panic while panicking". Even though using expect vs panic wouldn't address that iiuc, it's helpful to make our panics more informative.

illicitonion replied:
I agree that this is safe. If you really wanted to get rid of the `unwrap()` here, you could turn your `while` statement into a `loop` expression:
```rust
let error = loop {
    ...
    if num_retries == MAX_RETRIES || !is_retryable(&status) {
        break status;
    }
}

Err(error)
```

or even just use a `loop` which always returns from the function or continues to loop, rather than needing to store a `Option`. But I could certainly argue both sides of any of the approaches :)

Eric-Arellano replied:
`--remote-cache-max-retries` (or `--remote-store-max-retries`) does sound sensible to me. Not sure we need it, but makes sense.

tdyas replied:
I like that trick with the `loop` returning a value. Much cleaner.

tdyas commented:
`retry_call` will not work very well with the CAS streaming RPCs. It would probably be better to switch those to `BatchUpdateBlobs` since that is a single RPC call that can be wrapped easily (at least for small blobs).

Eric-Arellano reviewed (commented):
Thank you Tom! This is looking really good

illicitonion reviewed (commented):
Yeah, the approach looks very reasonable :)

tdyas commented:
I rebased the PR and made the following changes:

- Avoided using an `Option` in `retry_call`.
- Allow caller to pass in a function to evaluate whether an error is retriable or not.
- Made `retry_call` not rely on `tonic::Response` so it is generic over a `Result<T, E>`.
- Added a basic unit test to ensure it works.

tdyas commented:
This is ready for final review.

Eric-Arellano reviewed (approved):
Thanks! Would be good to CP to 2.5
